sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -546,33 +546,29 @@ case class MapConcat(children: Seq[Expression]) extends ComplexTypeMergingExpres
 
   override def nullable: Boolean = children.exists(_.nullable)
 
+  private lazy val mapBuilder = new ArrayBasedMapBuilder(dataType.keyType, dataType.valueType)
+
   override def eval(input: InternalRow): Any = {
-    val maps = children.map(_.eval(input))
+    val maps = children.map(_.eval(input).asInstanceOf[MapData]).toArray
```

Comment:
I need to access it by index below, turn it to array so that the access is guaranteed to be O(1).

Reply:
well, my understanding is that we could do a `maps.foreach` instead of accessing them by index. I don't see the index to be significant at all, but maybe I am missing something...

Reply:
in scala, while loop is faster than `foreach`. If you look at `Expression.eval` implementations, we use while loop a lot even `foreach` can produce simpler code.

Reply:
BTW, if it's not true anymore with scala 2.12, we should update them together with a bechmark, instead of only updating this single one.

Reply:
Yes, but converting `toArray` may require an extra O(N) operation for the copy, so I am not sure the difference between `while` and `foreach` is significant enough to cover the overhead of the copy...

Reply:
`toArray` is O(N) but we do it only once. If accessing by index is O(N), the total time complexity is O(N ^ 2).